### ISSUE TITLE
Control Replication Merge to Master

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#765]
See merge request:

https://gitlab.com/StanfordLegion/legion/-/merge_requests/157

TODO list:

- [x] Finish refactoring of instances for issue #546 
- [x] Refactor control replication to use the replicated instances for inline mappings and attach ops
- [x] Teach Legion Spy to validate inline mappings and attach operations in control replication
- [x] Detection of violations of control replication need to be improved
- [x] Verify semantics of attach/detach and inline mappings for control replication
- [x] Improve default mapper behavior for control replication (tracked in #896)
- [x] Fix implementation of `premap_task`
- [x] Ensure we have coverage of non-replicated Regent workloads in CI (@streichler )
- [x] Improve equivalence set refinement heuristics (#1309)

**Comment by lightsighter** [comment]
We can check off box 5 once we disable support for simultaneous coherence and relaxed coherence for inline mappings in control replicated contexts.

**Comment by lightsighter** [comment]
Instance refactoring towards the first bullet point is being tracked with this [merge request](https://gitlab.com/StanfordLegion/legion/-/merge_requests/183)

**Comment by lightsighter** [comment]
Most of the remains undone due to new equivalence set work. Moving to the next milestone.

**Comment by lightsighter** [comment]
Issue #546 has been completed and merged into the control replication branch. Probably the only thing left to do on the Legion side now is #1309 which is a proxy issue for fixing the equivalence set refinement heuristics.

**Comment by lightsighter** [comment]
Adding this to failing test to the requirements for the merge: http://sapling2.stanford.edu/~rupanshu/reproducer.cc

**Comment by lightsighter** [comment]
@rupanshusoi @elliottslaughter To handle the case you raised above, I've added support for control replication of tasks with region requirements in the `fixcrmap` branch (upstream from the `shardrefine` branch). It requires a backwards breaking change to the mapper interface for control replication which will need to be discussed at a Legion meeting. The interface is simpler than before and decouples the replication of tasks from mapping them, and is slightly backwards compatible through the default mapper (more on this in the meeting). There is a new mapper call named `replicate_task` which decides if and how to replicate a task, and then each of the copies (aka shards) will be mapped by `map_task` like normal. I've updated your example to use the new mapping interface (see attached). It runs in lots of different configurations with the `fixcrmap` branch. Note that it will also work if you mark the `reader` task as a leaf task.

[reproducer.tar.gz](https://github.com/StanfordLegion/legion/files/13807923/reproducer.tar.gz)
